src/util.ts

```diff
@@ -324,3 +326,86 @@ export const parsePemCertificateFromFile = async (path: string) => {
 
     return { cert: Certificate.fromBER(certDer), certPem, certDer };
 };
+
+type Device = {
+    /** A unique id that identifies the device to the platform tools and frida. */
+    id: string;
+    /** The name or model of the device (adb sometimes can not find that information easily, so it might be undefined) */
+    name?: string;
+    /** The platform the device is running on. */
+    platform: 'android' | 'ios';
+};
+
+/**
+ * Returns a list of all detected Android and iOS devices currently connected to the host. This includes Android
+ * emulators running on the host.
+ *
+ * @param options If the `frida` option is set to `true`, this function will use frida to detect the devices rather than
+ *   try to detect them using platform-specific tools (such as `adb` and `pymobiledevice3`).
+ */
+export const listDevices = async (options?: { frida?: boolean }): Promise<Device[]> => {
+    if (options?.frida)
+        return await frida
+            .enumerateDevices()
+            .then((devices) => devices.filter((device) => device.type === 'usb'))
+            .then((devices) =>
+                Promise.all(
+                    devices.map(async (device) => {
+                        const params = await device.querySystemParameters();
+                        return {
+                            id: device.id,
+                            name: device.name,
+                            platform: params.platform === 'darwin' ? 'ios' : 'android',
+                        };
+                    })
+                )
+            );
+
+    const venv = getVenv(venvOptions);
+    const python = async (...args: Parameters<Awaited<typeof venv>>) => (await venv)(...args);
+
+    const results = [];
```

Comment:
Why don't you make `aD` and `iD` the correct format already and just return `[...aD, ...iD]` instead of pushing to an array?

Reply:
Well, I liked how we could get all that information and wanted to implement a way to make them all available at first, but then I decided against it without changing the code. You are right.